Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (C) 2022-2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "Float16ArrayPixelBuffer.h"
+
+#if HAVE(HDR_SUPPORT)
+
+#include <JavaScriptCore/TypedArrayInlines.h>
+
+namespace WebCore {
+
+Ref<Float16ArrayPixelBuffer> Float16ArrayPixelBuffer::create(const PixelBufferFormat& format, const IntSize& size, JSC::Float16Array& data)
+{
+    return adoptRef(*new Float16ArrayPixelBuffer(format, size, { data }));
+}
+
+std::optional<Ref<Float16ArrayPixelBuffer>> Float16ArrayPixelBuffer::create(const PixelBufferFormat& format, const IntSize& size, std::span<const Float16> data)
+{
+    // FIXME: Support non-8 bit formats.
+    if (format.pixelFormat != PixelFormat::RGBA16F) {
+        ASSERT_NOT_REACHED();
+        return std::nullopt;
+    }
+
+    auto computedBufferSize = PixelBuffer::computeBufferSize(format.pixelFormat, size);
+    if (computedBufferSize.hasOverflowed()) {
+        ASSERT_NOT_REACHED();
+        return std::nullopt;
+    }
+
+    if (data.size_bytes() != computedBufferSize.value()) {
+        ASSERT_NOT_REACHED();
+        return std::nullopt;
+    }
+
+    auto buffer = JSC::Float16Array::tryCreate(data);
+    if (!buffer) {
+        ASSERT_NOT_REACHED();
+        return std::nullopt;
+    }
+
+    return Float16ArrayPixelBuffer::create(format, size, buffer.releaseNonNull());
+}
+
+RefPtr<Float16ArrayPixelBuffer> Float16ArrayPixelBuffer::tryCreate(const PixelBufferFormat& format, const IntSize& size)
+{
+    ASSERT(supportedPixelFormat(format.pixelFormat));
```

Comment:
This assertion doesn't seem right, since we want to check if the pixel format is supported by `Float16ArrayPixelBuffer` specifically. And similarly now in `ByteArrayPixelBuffer`, which we don't want to support `PixelFormat::RGBA16`.

Reply:
I'd like to keep this debug-time check, to make sure that the expected format here is supported by the overall `PixelBuffer::supportedPixelFormat()`.

But indeed we could also strengthen the parameter check to make sure we are called with exactly RGBA16F here.
(And similarly in the old ByteArray one.)
I was initially considering removing this parameter since it looks useless here, but decided to keep it (at least for now), in case present/future generic code really wants to provide a format.